src/main/scala/mesosphere/marathon/core/matcher/reconcile/impl/OfferMatcherReconciler.scala

```diff
@@ -37,60 +35,54 @@ private[reconcile] class OfferMatcherReconciler(instanceTracker: InstanceTracker
-    offer: Offer, resourcesByInstanceId: Map[Instance.Id, Seq[Resource]]): Future[MatchedInstanceOps] =
+  private[this] def processResourcesByReservationId(
+    offer: Offer, resourcesByReservationId: Map[Reservation.Id, Seq[Resource]]): Future[MatchedInstanceOps] =
     {
       // do not query instanceTracker in the common case
-      if (resourcesByInstanceId.isEmpty) Future.successful(MatchedInstanceOps.noMatch(offer.getId))
+      if (resourcesByReservationId.isEmpty) Future.successful(MatchedInstanceOps.noMatch(offer.getId))
       else {
-        def createInstanceOps(instancesBySpec: InstancesBySpec, rootGroup: RootGroup): MatchedInstanceOps = {
-
-          /* Was this task launched from a previous app definition, or a prior launch that did not clean up properly */
-          def spurious(instanceId: Instance.Id): Boolean =
-            instancesBySpec.instance(instanceId).isEmpty ||
-              (rootGroup.app(instanceId.runSpecId).isEmpty && rootGroup.pod(instanceId.runSpecId).isEmpty)
-
-          val instanceOps: Seq[InstanceOpWithSource] = resourcesByInstanceId.collect {
-            case (instanceId, spuriousResources) if spurious(instanceId) =>
-              val unreserveAndDestroy =
-                InstanceOp.UnreserveAndDestroyVolumes(
-                  stateOp = InstanceUpdateOperation.ForceExpunge(instanceId),
-                  oldInstance = instancesBySpec.instance(instanceId),
-                  resources = spuriousResources
-                )
-              logger.warn(s"removing spurious resources and volumes of $instanceId because the instance no longer exist")
-              InstanceOpWithSource(source(offer.getId), unreserveAndDestroy)
+        async {
+          val instancesBySpec = await(instanceTracker.instancesBySpec())
+          val knownReservationIds: Set[Reservation.Id] = instancesBySpec.allInstances.collect {
+            case Instance(_, _, _, _, _, Some(Reservation(_, _, reservationId))) => reservationId
           }(collection.breakOut)
 
-          MatchedInstanceOps(offer.getId, instanceOps, resendThisOffer = true)
+          createInstanceOps(offer, knownReservationIds, resourcesByReservationId)
         }
-
-        // query in parallel
-        val instancesBySpedFuture = instanceTracker.instancesBySpec()
-        val rootGroupFuture = groupRepository.root()
-
-        for { instancesBySpec <- instancesBySpedFuture; rootGroup <- rootGroupFuture }
-          yield createInstanceOps(instancesBySpec, rootGroup)
       }
     }
 
+  def createInstanceOps(offer: Offer, knownReservationIds: Set[Reservation.Id], resourcesByReservationId: Map[Reservation.Id, Seq[Resource]]): MatchedInstanceOps = {
+
+    val instanceOps: Seq[InstanceOpWithSource] = resourcesByReservationId.collect {
+      // TODO: Should we also unreserve terminal instances with goal == Decommission?
+      case (reservationId, spuriousResources) if !knownReservationIds.contains(reservationId) =>
```

Comment:
A comment in [InstanceUpdate](https://github.com/mesosphere/marathon/blob/master/src/main/scala/mesosphere/marathon/core/instance/update/InstanceUpdater.scala#L79) claims that we expunge an instance if the reservation is gone. However, `InstanceUpdater#shouldBeExpunged` does *not* check if the instance has a reservation. Thus we are left with reservations in Mesos which do not have an instance in Marathon.